### PR TITLE
Modernize shouldHaveUpperBound, shouldHaveLowerBound

### PR DESCRIPTION
**Pull request by okarmazin**
Continuation of work on #4354.

- Adds `shouldHaveUpper/LowerBound` for primitive arrays except BooleanArray. BooleanArray was left out in the interest of reducing API bloat. Booleans are better served with `shouldContain true/false`.
- Matcher messages report a more precise container type
- Adds more test cases for empty / single element containers
- Removes needless allocations: Iterables were being converted to lists, then checked iterated. Now they are just iterated once.

### CODE REVIEW

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt [hunk 1]
```diff
@@ -5,52 +5,253 @@ import io.kotest.matchers.Matcher
...
 }
 
-fun <T : Comparable<T>, C : Collection<T>> haveUpperBound(t: T) = object : Matcher<C> {
-   override fun test(value: C): MatcherResult {
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun <T : Comparable<T>> Array<T>.shouldHaveUpperBound(value: T): Array<T> {
```
Comment (LeoColman) [inline kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:87]: Do we need to have this overload? I think Iterable would catch it, right?

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt [hunk 2]
```diff
@@ -5,52 +5,253 @@ import io.kotest.matchers.Matcher
...
+   asList() should haveUpperBound(value, "Array")
+   return this
+}
+
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun <T : Comparable<T>, C : Collection<T>> C.shouldHaveUpperBound(value: T): C {
```
Comment (LeoColman) [inline kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:97]: Same for this one, Iterable should catch all

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt [hunk 3]
```diff
@@ -5,52 +5,253 @@ import io.kotest.matchers.Matcher
...
+   this should haveUpperBound(value, null)
+   return this
+}
+
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+fun <T : Comparable<T>, C : Collection<T>> haveUpperBound(value: T): Matcher<C> = haveUpperBound(value, null)
```
Comment (LeoColman) [inline kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:117]: Do we need C here?

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt [hunk 4]
```diff
@@ -5,52 +5,253 @@ import io.kotest.matchers.Matcher
...
+         { "$name should not have upper bound $t" })
+   }
+}
+
+private fun Iterable<*>.containerName(): String {
+   return when (this) {
+      is List -> "List"
+      is Set -> "Set"
+      is Map<*, *> -> "Map"
+      is ClosedRange<*>, is OpenEndRange<*> -> "Range"
```
Comment (LeoColman) [inline kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:135]: Maybe better to split these two for more information

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt [hunk 5]
```diff
@@ -5,52 +5,253 @@ import io.kotest.matchers.Matcher
...
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+fun <T : Comparable<T>, C : Collection<T>> haveUpperBound(value: T): Matcher<C> = haveUpperBound(value, null)
+
+private fun <T : Comparable<T>, I : Iterable<T>> haveUpperBound(t: T, name: String?): Matcher<I> = object : Matcher<I> {
+   override fun test(value: I): MatcherResult {
+      val name = name ?: value.containerName()
```
Comment (LeoColman) [inline kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:121]: As we are messing with this code already, perhaps we should guard against empty lists.

Empty lists will throw an error if they fail and pass silently otherwise.

We should place something like 'require(value.isNotEmpty())' or something similar. I think we do this in a few other collection matchers

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt [hunk 6]
```diff
@@ -5,52 +5,253 @@ import io.kotest.matchers.Matcher
...
+   asList() should haveLowerBound(value, "DoubleArray")
+   return this
+}
+
+/**
+ * Verifies that all elements are greater than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun <T : Comparable<T>> Array<T>.shouldHaveLowerBound(value: T): Array<T> {
```
Comment (LeoColman) [inline kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:221]: Same questioning from UpperBound: Should we have this overload?

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt [hunk 7]
```diff
@@ -5,52 +5,253 @@ import io.kotest.matchers.Matcher
...
+   asList() should haveLowerBound(value, "Array")
+   return this
+}
+
+/**
+ * Verifies that all elements are greater than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun <T : Comparable<T>, C : Collection<T>> C.shouldHaveLowerBound(value: T): C {
```
Comment (LeoColman) [inline kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:231]: Same here

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt [hunk 8]
```diff
@@ -5,52 +5,253 @@ import io.kotest.matchers.Matcher
...
+   this should haveLowerBound(value, null)
+   return this
+}
+
+/**
+ * Verifies that all elements are greater than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun <T : Comparable<T>, I : Iterable<T>> I.shouldHaveLowerBound(value: T): I {
```
Comment (LeoColman) [inline kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:241]: Do we actually need to define `I` here?

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt [hunk 9]
```diff
@@ -5,52 +5,253 @@ import io.kotest.matchers.Matcher
...
+   this should haveLowerBound(value, null)
+   return this
+}
+
+fun <T : Comparable<T>, C : Collection<T>> haveLowerBound(t: T): Matcher<C> = haveLowerBound(t, null)
+
+private fun <T : Comparable<T>, I : Iterable<T>> haveLowerBound(t: T, name: String?): Matcher<I> = object : Matcher<I> {
+   override fun test(value: I): MatcherResult {
+      val name = name ?: value.containerName()
```
Comment (LeoColman) [inline kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:250]: same comment here, we could add a neat empty-list guard here

**Review by LeoColman — CHANGES_REQUESTED** [review]
I believe we have some space to improve.
All in all, these are good modernizations :D

File: kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/collections/BoundsTest.kt [hunk 10]
```diff
@@ -10,24 +10,225 @@ class BoundsTest : WordSpec() {
...
+            longArrayOf(3) shouldHaveLowerBound 3
+            longArrayOf(1, 2, 3) shouldHaveLowerBound 1
+            longArrayOf(1, 2, 3) shouldHaveLowerBound 0
+
+            floatArrayOf() shouldHaveLowerBound 0.3f
+            floatArrayOf(0.3f) shouldHaveLowerBound 0.3f
+            floatArrayOf(0.1f, 0.2f, 0.3f) shouldHaveLowerBound 0.1f
+            floatArrayOf(0.1f, 0.2f, 0.3f) shouldHaveLowerBound 0f
+
+            doubleArrayOf() shouldHaveLowerBound 0.3
```
Comment (LeoColman) [inline kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/collections/BoundsTest.kt:155]: As an example for the aforementioned guard-clauses, an empty double array will not have 0.3 as lowerbound

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt (hunk 2 above)
Comment (okarmazin) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:97]: Yes, a followup PR is already waiting - https://github.com/kotest/kotest/pull/4379

As I said before, I was instructed by @AlexCue987 to make these removals into separate PRs.

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt (hunk 1 above)
Comment (okarmazin) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:87]: > I think Iterable would catch it, right?

No, `Array` does not implement `Iterable` - `Array<MyType>` is just `MyType[]`.

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt (hunk 4 above)
Comment (okarmazin) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:135]: https://github.com/kotest/kotest/issues/4381

Will resolve separately

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt (hunk 5 above)
Comment (okarmazin) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:121]: This is up for debate IMO. If you word `hasUpperBound(T)` as "all elements are less than or equal to T", then an empty set does evaluate to true, because there are no elements to violate the constraint. This is known as a [_vacuous truth_](https://en.wikipedia.org/wiki/Vacuous_truth) -- statement is true, but doesn't really say anything.

I argue current behavior is correct. If you want to check for presence of elements, there is `shouldNotBeEmpty` for that.

See also: [stdlib function `all`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/all.html)

@AlexCue987 @sksamuel What do you think?

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt (hunk 1 above)
Comment (okarmazin) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:87]: In general a collection assertion has to have these overloads to cover all "reasonable" receivers:

1) primitive arrays
2) `Array<T>`
3) `Iterable<T>` (also covers all Collections and some Ranges)
3) `Sequence<T>` (Sequence does not implement `Iterable`, but can give you an iterator - cheaply representable as `Iterable`. Big opportunity for reusing the Iterable Matcher implementation under the hood.)
4) If appropriate, Map<K, T> - if the assertion makes sense to treat Map as a collection of Entries. Otherwise, Map doesn't implement Iterable on its own.


All of these can share a single `Matcher<Iterable<T>>` implementation since they are either Iterables or can be cheaply represented as an Iterable without big allocations (`Array.asList`, `Sequence.asIterable`).

File: kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/collections/BoundsTest.kt (hunk 10 above)
Comment (okarmazin) [reply kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/collections/BoundsTest.kt:155]: Yep, I'll close this comment as resolved to keep that discussion in one thread - https://github.com/kotest/kotest/pull/4378#discussion_r1779820867

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt (hunk 9 above)
Comment (okarmazin) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:250]: Yep, I'll close this comment as resolved to keep that discussion in one thread - https://github.com/kotest/kotest/pull/4378#discussion_r1779820867

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt (hunk 7 above)
Comment (okarmazin) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:231]: Yep, scheduled for already in #4379

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt (hunk 8 above)
Comment (okarmazin) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:241]: Someone requested the underlying type to be returned due to having custom Iterable implementations. That way he could preserve the underlying type through assertion chain.

See https://github.com/kotest/kotest/issues/3960

Comment (okarmazin) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:241]: As returning the underlying type is the currently accepted behavior - and was so before this PR - I'll close this as resolved. Please reopen if you think there should be more discussion or if you think this should be changed to return Iterable again. But keep in mind that someone requested this behavior and it was accepted.

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt (hunk 3 above)
Comment (okarmazin) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:117]: Nope, scheduled for removal #4379

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt (hunk 5 above)
Comment (okarmazin) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:121]: > Empty lists will throw an error if they fail and pass silently otherwise.

I apologize, I don't understand this sentence. Could you please reword it?

**Comment by okarmazin** [comment]
More context for this PR: https://github.com/kotest/kotest/pull/4369#issuecomment-2381280124

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt (hunk 5 above)
Comment (LeoColman) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:121]: > I argue current behavior is correct. If you want to check for presence of elements, there is shouldNotBeEmpty for that.

Exactly. If we want to check for empty collections it should happen on `shouldNotBeEmpty`, and not accidentally on `shouldHaveUpperBound`. Therefore, if you're making a check against a collection but it's actually empty, we should prevent such thing to happen

> statement is true, but doesn't really say anything.

So the test would basically do nothing? That's not a very useful test, is it?



https://github.com/kotest/kotest/blob/1360aa5a85e5e44bdf85ef04718ad24019ca47bf/kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/ranges/bein.kt#L54

https://github.com/kotest/kotest/blob/1360aa5a85e5e44bdf85ef04718ad24019ca47bf/kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/matchers.kt#L133

These are examples in our codebase when we do this kind of guard clause.


> See also: [stdlib function all](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/all.html)

The stdlib function is for general purpose, and not testing/assertion purposes. I think this comparison doesn't apply

Comment (okarmazin) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:121]: I don't have a horse in this race. There's just a decision to be made about which behavior is deemed more in line with the principle of least surprise. 
I argued that I'd like to follow mathematical logic - but I'll happily throw an error on empty container. It just must be consistent everywhere. That's the important bit.

> So the test would basically do nothing? That's not a very useful test, is it?

No, it doesn't "do nothing". It verifies that the statement is true, which it is. An assertion is just a logical statement. Now, is it desirable for `hasLowerBound` to be just `hasLowerBound`, or should it be a conjunction of `isNotEmpty && hasLowerBound`?

Comment (okarmazin) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:121]: It is also important to note that this PR doesn't modify previous behavior. It would be a behavioral change if this proposed `isNotEmpty` condition was added.

If all collection assertions should suddenly start rejecting empty containers, that's a separate issue for a separate PR. I'll open a ticket for it.

Comment (LeoColman) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt:121]: > It is also important to note that this PR doesn't modify previous behavior. It would be a behavioral change if this proposed isNotEmpty condition was added.

Fair point, I'll resolve this as we can track it in the new ticket you created

**Review by LeoColman — APPROVED** [review]
